android/java/org/chromium/chrome/browser/app/domain/MarketModel.java

```diff
@@ -0,0 +1,95 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+package org.chromium.chrome.browser.app.domain;
+
+import androidx.annotation.NonNull;
+
+import org.chromium.brave_wallet.mojom.AssetRatioService;
+import org.chromium.brave_wallet.mojom.CoinMarket;
+
+import java.math.RoundingMode;
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
+import java.util.Locale;
+
+public class MarketModel {
+    private static final String CURRENCY = "USD";
+    // Fetch the first 250 assets.
+    private static final byte ASSETS_REQUEST_LIMIT = (byte) (250 & 0xFF);
+    private AssetRatioService mAssetRatioService;
+
+    private final DecimalFormatSymbols mSymbols;
+    private final DecimalFormat mPriceFormatter;
+    private final DecimalFormat mPercentageFormatter;
+
+    public MarketModel(AssetRatioService assetRatioService) {
+        mAssetRatioService = assetRatioService;
+
+        mSymbols = new DecimalFormatSymbols(Locale.ENGLISH);
+        mPriceFormatter = new DecimalFormat("$0.00##########", mSymbols);
+        mPriceFormatter.setGroupingSize(3);
+        mPriceFormatter.setGroupingUsed(true);
+
+        mPercentageFormatter = new DecimalFormat("0.00", mSymbols);
+        mPercentageFormatter.setRoundingMode(RoundingMode.UP);
+    }
+
+    public void resetService(AssetRatioService assetRatioService) {
+        mAssetRatioService = assetRatioService;
+    }
+
+    /**
+     * Gets the first 250 market assets.
+     * @param callback Callback used to notify when the assets have been downloaded.
+     */
+    public void getCoinMarkets(@NonNull CoinMarketsCallback callback) {
+        if (callback == null) {
+            throw new IllegalArgumentException("Callback must not be null.");
+        }
+        if (mAssetRatioService == null) {
+            return;
+        }
+        mAssetRatioService.getCoinMarkets(
+                CURRENCY, ASSETS_REQUEST_LIMIT, (success, coinMarkets) -> {
+                    if (success) {
+                        callback.onCoinMarketsSuccess(coinMarkets);
+                    } else {
+                        callback.onCoinMarketsFail();
+                    }
+                });
+    }
```

Comment:
use live data instead of custom callback.

Reply:
Okay 👍
What's the advantage of using live data instead of callback here?

Reply:
Livedata ensures safer UI updates and it prevent crashes.
[if its lifecycle is in the STARTED or RESUMED state. LiveData only notifies active observers about updates. Inactive observers registered to watch LiveData objects aren't notified about changes.](https://developer.android.com/topic/libraries/architecture/livedata)

Reply:
Introducing live data caused some pain because of this: https://issuetracker.google.com/issues/37007605#hc141

Eventually, adding `notifyItemRangeRemoved(0, previousContentSize)` before `notifyItemRangeInserted(0, newContentSize)` fixed all the issues.